### PR TITLE
Add new Array Coerce Cast Metadata object

### PR DESCRIPTION
**Pull request by khannaekta**
Currently executor crashes for following query with ORCA ON :
```
CREATE TABLE FOO(a
integer NOT NULL, b double precision[]); SELECT b FROM foo UNION ALL SELECT
ARRAY[90, 90] as Cont_features;
```

In this query, we are appending an integer array (ARRAY[90, 90]) to a double
precision array (foo.b) and hence we need to apply a cast on ARRAY[90, 90] to
generate ARRAY[90, 90]::double precision[].
In gpdb5 there is not direct function available that can cast array of any type
to array of any other type. So in relcache to dxl translator we look into the
array elements and get their type and try to find a cast function for them.
For this query, source type is 23 (integer) and dest type is 701 (double precision)
 and we try to find if we have a conversion function for 23 -> 701.
Since that is available we send that function to ORCA as follows:
```
<dxl:MDCast
Mdid="3.1007.1.0;1022.1.0" Name="float8" BinaryCoercible="false"
SourceTypeId="0.1007.1.0" DestinationTypeId="0.1022.1.0"
CastFuncId="0.316.1.0"/>
```
Here we are misinforming ORCA by specifying that function with id 316 is available
to convert type 1007 (integer array) to 1022 (double precision) array.
However Function id 316 is simple int4 to float8 conversion function and it CAN NOT
convert an array of int4 to array of double precision.
ORCA generates a plan using this function but executor crashes while executing this
function because this function can not handle arrays.

This commit adds a new ArrayCoerceCast MetaData object which will be constructed
when we need to convert an array of one type to another; instead of constructing
Cast Metadata.

`CMDArrayCoerceCastGPDB` extends `CMDCastGPDB` in that it includes necessary information
to generate `CScalarArrayCoerceExpr`.

In Relcache Translator on GPDB, we will construct an object of `CMDArrayCoerceCastGPDB`
instead of `CMDCastGPDB` depending on the coercion path determined by `FCastFunc`.

Added relevant test cases.

### CODE REVIEW

File: data/dxl/parse_tests/q26-Metadata.xml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
   <dxl:Metadata>
-    <dxl:MDCast Mdid="3.23.1.0;20.1.0" Name="int8" BinaryCoercible="false" SourceTypeId="0.23.1.0" DestinationTypeId="0.20.1.0" CastFuncId="0.481.1.0"/>
-    <dxl:MDCast Mdid="3.23.1.0;26.1.0" Name="int" BinaryCoercible="true" SourceTypeId="0.23.1.0" DestinationTypeId="0.26.1.0" CastFuncId="0.0.0.0"/>
+    <dxl:MDCast Mdid="3.23.1.0;20.1.0" Name="int8" BinaryCoercible="false" SourceTypeId="0.23.1.0" DestinationTypeId="0.20.1.0" CastFuncId="0.481.1.0" CoercePathType="1"/>
+    <dxl:MDCast Mdid="3.23.1.0;26.1.0" Name="int" BinaryCoercible="true" SourceTypeId="0.23.1.0" DestinationTypeId="0.26.1.0" CastFuncId="0.0.0.0" CoercePathType="0"/>
```
Comment (vraghavan78) [inline data/dxl/parse_tests/q26-Metadata.xml]: 1 and 0 does not read well. Can we have a string?

File: libgpopt/src/translate/CTranslatorDXLToExpr.cpp [hunk 2]
```diff
@@ -724,14 +725,29 @@ CTranslatorDXLToExpr::PexprCastPrjElem
...
+		pexprCast =
 		GPOS_NEW(m_pmp) CExpression
-			(
+		(
 			m_pmp,
-			GPOS_NEW(m_pmp) CScalarCast(m_pmp, pmdidDest, pmdcast->PmdidCastFunc(), pmdcast->FBinaryCoercible()),
+			GPOS_NEW(m_pmp) CScalarArrayCoerceExpr(m_pmp, parrayCoerceCast->PmdidCastFunc(), pmdidDest, parrayCoerceCast->IMod(), parrayCoerceCast->FIsExplicit(), (COperator::ECoercionForm) parrayCoerceCast->Ecf(), parrayCoerceCast->ILoc()),
 			GPOS_NEW(m_pmp) CExpression(m_pmp, GPOS_NEW(m_pmp) CScalarIdent(m_pmp, pcrToCast))
+		);
+	}
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorDXLToExpr.cpp]: why are not handling EmdtCoerceViaIO. Either handle it or the kill the enum

File: data/dxl/parse_tests/q26-Metadata.xml (hunk 1 above)
Comment (khannaekta) [inline data/dxl/parse_tests/q26-Metadata.xml]: Sounds like a good idea however the dxl entry for `CScalarArrayCoerceExpr` does not print the string either:
```
<dxl:ArrayCoerceExpr ElementFunc="0.0.0.0" TypeMdid="0.1009.1.0" TypeModification="-1" IsExplicit="false" CoercionForm="2" Location="0">
```
We will be inconsistent with this. Thoughts?

File: libnaucrates/src/md/CMDArrayCoerceCastGPDB.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+#include "naucrates/dxl/xml/CXMLSerializer.h"
+#include "naucrates/dxl/CDXLUtils.h"
+
+using namespace gpmd;
+using namespace gpdxl;
+
+// ctor
+CMDArrayCoerceCastGPDB::CMDArrayCoerceCastGPDB
+	(
+	 IMemoryPool *pmp,
```
Comment (oarap) [inline libnaucrates/src/md/CMDArrayCoerceCastGPDB.cpp]: remove extra space  before the parameters. <tab><space>ImemoryPool ... . Better to remove that space

File: libnaucrates/src/parser/CParseHandlerFactory.cpp [hunk 4]
```diff
@@ -3596,4 +3597,16 @@ CParseHandlerFactory::PphValuesScan
 	return GPOS_NEW(pmp) CParseHandlerValuesScan(pmp, pphm, pphRoot);
 }
 
+// creates a parse handler for parsing GPDB-specific array coerce cast metadata
+CParseHandlerBase *
+CParseHandlerFactory::PphMDArrayCoerceCast
+(
```
Comment (oarap) [inline libnaucrates/src/parser/CParseHandlerFactory.cpp]: indent the (

File: libnaucrates/src/parser/CParseHandlerMDArrayCoerceCast.cpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	BOOL fBinaryCoercible = CDXLOperatorFactory::FValueFromAttrs
+											(
+											m_pphm->Pmm(),
+											attrs,
+											EdxltokenGPDBCastBinaryCoercible,
+											EdxltokenGPDBArrayCoerceCast
+											);
+
+	IMDCast::EmdCoercepathType eCoercePathType = (IMDCast::EmdCoercepathType) CDXLOperatorFactory::IValueFromAttrs
+																							(
```
Comment (oarap) [inline libnaucrates/src/parser/CParseHandlerMDArrayCoerceCast.cpp]: something is off on here for the spacing

File: libnaucrates/src/parser/CParseHandlerMDArrayCoerceCast.cpp [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	BOOL fIsExplicit =CDXLOperatorFactory::FValueFromAttrs
+									(
+									 m_pphm->Pmm(),
+									 attrs,
+									 EdxltokenIsExplicit,
+									 EdxltokenGPDBArrayCoerceCast
+									 );
+
+	EdxlCoercionForm edcf = (EdxlCoercionForm) CDXLOperatorFactory::IValueFromAttrs
+																		(
```
Comment (oarap) [inline libnaucrates/src/parser/CParseHandlerMDArrayCoerceCast.cpp]: extra space which Xcode puts, in every parameter.

File: libnaucrates/src/parser/CParseHandlerMDCast.cpp [hunk 7]
```diff
@@ -117,8 +117,17 @@ CParseHandlerMDCast::StartElement
 											EdxltokenGPDBCastBinaryCoercible,
 											EdxltokenGPDBCast
 											);
-	
-	m_pimdobj = GPOS_NEW(m_pmp) CMDCastGPDB(m_pmp, pmdid, pmdname, pmdidSrc, pmdidDest, fBinaryCoercible, pmdidCastFunc);
+
+	IMDCast::EmdCoercepathType eCoercePathType = (IMDCast::EmdCoercepathType) CDXLOperatorFactory::IValueFromAttrs
+																							(
```
Comment (oarap) [inline libnaucrates/src/parser/CParseHandlerMDCast.cpp]: something is off as well on here.

File: libnaucrates/include/naucrates/md/IMDCast.h [hunk 8]
```diff
@@ -35,6 +35,16 @@ namespace gpmd
 	{	
 		public:
 
+			// type of coercion pathway
+			enum EmdCoercepathType
+			{
+				EmdtNone,			/* failed to find any coercion pathway */
+				EmdtFunc,			/* apply the specified coercion function */
+				EmdtRelableType,	/* binary-compatible cast, no function */
```
Comment (hsyuan) [inline libnaucrates/include/naucrates/md/IMDCast.h]: EmdtRelableType -> EmdtRelabelType

File: libgpopt/src/translate/CTranslatorDXLToExpr.cpp (hunk 2 above)
Comment (khannaekta) [inline libgpopt/src/translate/CTranslatorDXLToExpr.cpp]: Removed the Enum, as we will not be receiving this in the coerce path.

**Comment by vraghavan78** [comment]
The following functions need to be looked at before we can consider merging.

```
// cast the input column reference to the destination mdid
CExpression *
CUtils::PexprCast
	(
	IMemoryPool *pmp,
	CMDAccessor *pmda,
	const CColRef *pcr,
	IMDId *pmdidDest
	)


// add explicit casting on the input expression to the destination type
CExpression *
CPredicateUtils::PexprCast
	(
	IMemoryPool *pmp,
	CMDAccessor *pmda,
	CExpression *pexpr,
	IMDId *pmdidDest
	)


BOOL
CTranslatorDXLToExprUtils::FCastFunc
	(
	CMDAccessor *pmda,
	const CDXLNode *pdxln,
	IMDId *pmdidInput
	)

CDXLNode *
CTranslatorExprToDXL::PdxlnScCmpPartKey
	(
	CExpression *pexprScCmp,
	CColRef *pcrPartKey,
	IMDId *pmdidTypePartKey,
	ULONG ulPartLevel,
	BOOL fRangePart,
	BOOL *pfLTComparison,	// input/output
	BOOL *pfGTComparison,	// input/output
	BOOL *pfEQComparison	// input/output
	)
```

**Comment by vraghavan78** [comment]
Also, I would suggest we pull in all cast functions into a separate file CUtilsCasts.cpp or something and add the following functions there:

CUtils
* CExpression *PexprCast(IMemoryPool *pmp, CMDAccessor *pmda, CExpression *pexpr, IMDId *pmdidDest);
* CExpression *PexprCast( IMemoryPool *pmp, CMDAccessor *pmda, const CColRef *pcr, IMDId *pmdidDest);
* BOOL FBinaryCoercibleCast(CExpression *pexpr);
*  BOOL FCmpOrCastedCmpExists(IMDId *pmdidLeft, IMDId *pmdidRight, IMDType::ECmpType ecmpt);
* BOOL FBinaryCoercibleCastedScId(CExpression *pexpr, CColRef *pcr);
* BOOL FBinaryCoercibleCastedScId(CExpression *pexpr);
* CExpression *PexprWithoutBinaryCoercibleCasts(CExpression *pexpr);
* const CColRef *PcrExtractFromScIdOrCastScId(CExpression *pexpr);

CPredicateUtils.cpp
* CExpression *PexprAddCast(IMemoryPool *pmp, CExpression *pexprPred);
* DrgPexpr *PdrgpexprCastEquality(IMemoryPool *pmp, CExpression *pexpr);
* CExpression *PexprCast(IMemoryPool *pmp, CMDAccessor *pmda, CExpression *pexpr, IMDId *pmdidDest);

File: libgpopt/src/translate/CTranslatorDXLToExpr.cpp (hunk 2 above)
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorDXLToExpr.cpp]: This function should go in the CUtilsCasts.cpp file that you should create

File: libgpopt/src/translate/CTranslatorDXLToExpr.cpp [hunk 9]
```diff
@@ -3572,12 +3606,37 @@ CTranslatorDXLToExpr::PexprScalarCast
 	const IMDCast *pmdcast = m_pmda->Pmdcast(pmdidInput, pmdidType);
 	BOOL fRelabel = pmdcast->FBinaryCoercible();
 
-	CExpression *pexpr = GPOS_NEW(m_pmp) CExpression
+	CExpression *pexpr;
+	
+	if(pmdcast->EmdPathType() == IMDCast::EmdtArrayCoerce)
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorDXLToExpr.cpp]: if (

**Pull request merged**